lib/rpush/client/active_model/fcm/notification.rb

```diff
@@ -0,0 +1,83 @@
+module Rpush
+  module Client
+    module ActiveModel
+      module Fcm
+        module Notification
+          FCM_PRIORITY_HIGH = Rpush::Client::ActiveModel::Apns::Notification::APNS_PRIORITY_IMMEDIATE
+          FCM_PRIORITY_NORMAL = Rpush::Client::ActiveModel::Apns::Notification::APNS_PRIORITY_CONSERVE_POWER
+          FCM_PRIORITIES = [FCM_PRIORITY_HIGH, FCM_PRIORITY_NORMAL]
+
+          def self.included(base)
+            base.instance_eval do
+              validates :device_token, presence: true
+              validates :priority, inclusion: { in: FCM_PRIORITIES }, allow_nil: true
+
+              validates_with Rpush::Client::ActiveModel::PayloadDataSizeValidator, limit: 4096
+              validates_with Rpush::Client::ActiveModel::RegistrationIdsCountValidator, limit: 1000
+
+              validates_with Rpush::Client::ActiveModel::Fcm::ExpiryCollapseKeyMutualInclusionValidator
+            end
+          end
+
+          # This is a hack. The schema defines `priority` to be an integer, but FCM expects a string.
+          # But for users of rpush to have an API they might expect (setting priority to `high`, not 10)
+          # we do a little conversion here.
+          # I'm not happy about it, but this will have to do until I can take a further look.
+          def priority=(priority)
+            case priority
+              when 'high', FCM_PRIORITY_HIGH
+                super(FCM_PRIORITY_HIGH)
+              when 'normal', FCM_PRIORITY_NORMAL
+                super(FCM_PRIORITY_NORMAL)
+              else
+                errors.add(:priority, 'must be one of either "normal" or "high"')
+            end
+          end
+
+          def as_json(options = nil) # rubocop:disable Metrics/PerceivedComplexity
+            json = {
+              'data' => data,
+              'android' => android_config,
+              'token' => device_token
+            }
+            json['content_available'] = content_available if content_available
+            json['notification'] = alert if alert
```

Comment:
@mirkode 

Sorry for jumping in. Thank you for implementing FCM support.

I'm trying to send push notification to iOS app with this branch, but the notification does not show up. 

```ruby
n = Rpush::Fcm::Notification.new

n.app = Rpush::Fcm::App.find_by(name: '...')
n.device_token = '***'
n.priority = 'normal'
n.notification = { title: 'Hello' }

n.save!
```

Rpush log says it sent succesfully but the `Notification` JSON does not have `message.notification` field.

```
pry(main)> Rpush::Fcm::Notification.first.as_json
=> {"message"=>
  {"data"=>nil,
   "android"=>
    {"notification"=>
      {"title"=>"Hello", "notification_priority"=>"PRIORITY_DEFAULT", "default_sound"=>true},
     "priority"=>5,
     "ttl"=>"86400s"},
   "token"=> "***"}}
```

Then I dig into this PR code, found this line. I am not sure this is the cause of the problem I have, I guess this should be:

```suggestion
            json['notification'] = notification if notification
```

or handle `alert` like https://github.com/rpush/rpush/blob/31ae99cf4214c2e51d5c7771742558b4cf2e74a4/lib/rpush/client/active_record/apns/active_record_serializable_notification.rb#L16-L32

Reply:
Hi @Joe-noh ,

Thanks for your suggestion! 
If it is working for you, I don't see any issue changing it to that.
I think I remember that I had a reason to change it to that, but I would need to look into this again, since this was quite some time ago :)

Thanks a lot

Reply:
@mirkode

I forked and committed https://github.com/Joe-noh/rpush/commit/9c81ece8d6f916bee230ef10560ae1000dcab940, then successfully pushed notification to iOS app. A bit naive implementation though, just FYI.